Kubernetes Access Logs - Final Field In Config Not Dropped

### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

<!--

HOW TO WRITE A GOOD BUG REPORT?

- Respect the issue template as much as possible.
- If possible, use the command `traefik bug`. See https://www.youtube.com/watch?v=Lyz62L8m93I.
- The title should be short and descriptive.
- Explain the conditions which led you to report this issue: the context.
- The context should lead to something, an idea or a problem that you’re facing.
- Remain clear and concise.
- Format your messages to help the reader focus on what matters and understand the structure of your message, use Markdown syntax https://help.github.com/articles/github-flavored-markdown

-->

I enabled access logging on my Tæfik ingress controller. I do not want to log `Username`, `Hostname`, `ClientUsername` and `ClientHost`.

To do this, I passed the command line arguments:

```console
--accessLog.fields.defaultMode="keep"
--accessLog.fields.names="Username=drop Hostname=drop ClientUsername=drop ClientHost=drop"
```

The same thing occurs with headers:

```console
--accessLog.fields.headers.defaultMode=keep
--accessLog.fields.headers.names="Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact"
```

### What did you expect to see?

I expected `ClientHost` to not appear in the logs.

I expected to see `"request_X-Forwarded-For":"REDACTED"` in the JSON logs.

### What did you see instead?

`ClientHost` was present in the access logs.
`request_X-Forwarded-For` had an IP address.

All other fields were dropped and redacted.

### Output of `traefik version`: (_What version of Traefik are you using?_)

Noticed the problem on `1.6.4`

Recreated the issue on `1.6.6`

Recreated the issue on `v1.7.0-rc3`

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

The environment is a Kops cluster in AWS running `v1.9.7` of Kubernetes.

I'm not using a TOML file, just the command line args in a Kubernetes `yaml` file.

Below is my `yaml` file for the deployment.

```yaml
---
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: traefik-ingress-lb
  labels:
    app: traefik-ingress-lb
    tier: proxy
spec:
  replicas: 2
  selector:
    matchLabels:
      app: traefik-ingress-lb
      tier: proxy
  template:
    metadata:
      labels:
        app: traefik-ingress-lb
        tier: proxy
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
        scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
    spec:
      serviceAccountName: traefik-ingress
      terminationGracePeriodSeconds: 60
      containers:
      - image: traefik:v1.7
        name: traefik-ingress-lb

        ports:
        - name: http
          containerPort: 80
        - name: https
          containerPort: 443
        - name: admin
          containerPort: 8080

        args:
        - --traefikLog.format=json
        - --accessLog.format=json
        - --accessLog.fields.defaultMode="keep"
        - --accessLog.fields.names="Username=drop Hostname=drop ClientUsername=drop ClientHost=drop"
        - --accessLog.fields.headers.defaultMode=keep
        - --accessLog.fields.headers.names="Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact"
        - --api
        - --api.dashboard
        - --api.statistics
        - --kubernetes
        - --kubernetes.ingressClass=traefik-external
        - --kubernetes.namespaces=app
        - --metrics.prometheus
        - --logLevel=INFO
        - --entryPoints=Name:http Address::80
        - --entryPoints=Name:https Address::443 TLS
        - --defaultentrypoints=http

        resources:
          requests:
            cpu: "100m"
            memory: "20Mi"
          limits:
            cpu: "150m"
            memory: "40Mi"

        readinessProbe:
          tcpSocket:
            port: 80
          failureThreshold: 1
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
        livenessProbe:
          tcpSocket:
            port: 80
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
```

#### Logs
Each request makes an access log for the pod, containing

`"request_X-Forwarded-For":"12.34.45.67"` (Should be REDACTED)
`"ClientHost":"12.34.56.78"` (Shouldn't be here)

### Fix

A dirty fix is to add an extra field to each command line arg:

```console
--accessLog.fields.names="Username=drop Hostname=drop ClientUsername=drop ClientHost=drop FakeField=drop"
--accessLog.fields.headers.names="Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact X-Clacks-Overhead=keep"
```

This then makes the logs function as expected.

### Recreation

I can only re-create this bug with the exact `yaml` file I put above.

I tried just using `docker run`  with the same args. The log fields were removed and redacted as expected.

I tried running træfik with the same args by itself in `minikube` cluster on my laptop and again, the fields were dropped and redacted.

I think I see what the problem is, and I'm going to try to fix it. In the meantime, a workaround that works for me is to simply remove the double quotes like this:

```
        - --accessLog.fields.names=Username=drop Hostname=drop ClientUsername=drop ClientHost=drop
        - --accessLog.fields.headers.names=Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact
```

It seems the issue is that because of the YAML format, the parser gets escaped characters when there are double quotes in the line, and that's what is making it ignore the last key-value pair.

Also, I found another issue while experimenting with this, which is that if we add a space at the end without providing another key/value pair, the parser crashes:

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/containous/traefik/types.(*FieldHeaderNames).Set(0xc0000a6020, 0xc0001f00b1, 0x61, 0x1e, 0x2)
	/go/src/github.com/containous/traefik/types/logs.go:130 +0x153
github.com/containous/traefik/vendor/github.com/ogier/pflag.(*FlagSet).setFlag(0xc0006a7280, 0xc000357040, 0xc0001f00b1, 0x61, 0xc0001f0090, 0x82, 0x2, 0x2)
	/go/src/github.com/containous/traefik/vendor/github.com/ogier/pflag/flag.go:464 +0x63
github.com/containous/traefik/vendor/github.com/ogier/pflag.(*FlagSet).parseArgs(0xc0006a7280, 0xc0000ac4e0, 0xc, 0xb, 0x0, 0x11)
	/go/src/github.com/containous/traefik/vendor/github.com/ogier/pflag/flag.go:516 +0xc28
github.com/containous/traefik/vendor/github.com/ogier/pflag.(*FlagSet).Parse(0xc0006a7280, 0xc0000ac480, 0x11, 0x11, 0x11, 0x11)
	/go/src/github.com/containous/traefik/vendor/github.com/ogier/pflag/flag.go:563 +0xa1
github.com/containous/traefik/vendor/github.com/containous/flaeg.parseArgs(0xc0000ac010, 0x11, 0x11, 0xc0006a7500, 0xc0005dc060, 0x16, 0xc0006a74d0, 0x0)
	/go/src/github.com/containous/traefik/vendor/github.com/containous/flaeg/flaeg.go:139 +0x7e9
github.com/containous/traefik/vendor/github.com/containous/flaeg.LoadWithCommand(0xc0002ba120, 0xc0000ac010, 0x11, 0x11, 0xc0002a9800, 0xc0005ba140, 0x5, 0x8, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/github.com/containous/flaeg/flaeg.go:419 +0x35b
github.com/containous/traefik/vendor/github.com/containous/flaeg.(*Flaeg).Parse(0xc0002ba1e0, 0xc0002ba120, 0x0, 0x0, 0x5)
	/go/src/github.com/containous/traefik/vendor/github.com/containous/flaeg/flaeg.go:640 +0x7d
main.main()
	/go/src/github.com/containous/traefik/cmd/traefik/traefik.go:91 +0x1401
```

--- a/types/logs.go
+++ b/types/logs.go
@@ -88,6 +88,11 @@ func (f *FieldNames) Get() interface{} {
 // Set's argument is a string to be parsed to set the flag.
 // It's a space-separated list, so we split it.
 func (f *FieldNames) Set(value string) error {
+	// When arguments are passed through YAML, escaped double quotes
+	// might be added to this string, and they would break the last
+	// key/value pair. This ensures the string is clean.
+	value = strings.Replace(value, "\"", "", -1)
+
 	fields := strings.Fields(value)
 
 	for _, field := range fields {
@@ -123,6 +128,11 @@ func (f *FieldHeaderNames) Get() interface{} {
 // Set's argument is a string to be parsed to set the flag.
 // It's a space-separated list, so we split it.
 func (f *FieldHeaderNames) Set(value string) error {
+	// When arguments are passed through YAML, escaped double quotes
+	// might be added to this string, and they would break the last
+	// key/value pair. This ensures the string is clean.
+	value = strings.Replace(value, "\"", "", -1)
+
 	fields := strings.Fields(value)
 
 	for _, field := range fields {

--- a/types/logs.go
+++ b/types/logs.go
@@ -91,7 +91,7 @@ func (f *FieldNames) Set(value string) error {
 	// When arguments are passed through YAML, escaped double quotes
 	// might be added to this string, and they would break the last
 	// key/value pair. This ensures the string is clean.
-	value = strings.Replace(value, "\"", "", -1)
+	value = strings.Trim(value, "\"")
 
 	fields := strings.Fields(value)
 
@@ -131,7 +131,7 @@ func (f *FieldHeaderNames) Set(value string) error {
 	// When arguments are passed through YAML, escaped double quotes
 	// might be added to this string, and they would break the last
 	// key/value pair. This ensures the string is clean.
-	value = strings.Replace(value, "\"", "", -1)
+	value = strings.Trim(value, "\"")
 
 	fields := strings.Fields(value)
 

--- a/types/logs_test.go
+++ b/types/logs_test.go
@@ -301,6 +301,14 @@ func TestFieldsHeadersNamesSet(t *testing.T) {
 				"X-HEADER-2": "bar",
 			},
 		},
+		{
+			desc:  "Two values separated by space with escaped double quotes should return FieldNames of size 2",
+			value: "\"X-HEADER-1=foo X-HEADER-2=bar\"",
+			expected: &FieldHeaderNames{
+				"X-HEADER-1": "foo",
+				"X-HEADER-2": "bar",
+			},
+		},
 	}
 
 	for _, test := range testCases {

